### PR TITLE
Part 1

### PR DESCRIPTION
**Pull request by JaneDeng28**

### CODE REVIEW

**Comment by JaneDeng28** [comment]
I tried rebase but I am not sure if it works

File: ehr_analysis2.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+from datetime import date, timedelta, datetime
+from typing import Union
+
+"""Data Parsing"""
+def parse_data(filename):
```
Comment (patrickkwang) [inline ehr_analysis2.py:5]: Need type hints for all function inputs and outputs.

File: ehr_analysis2.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                key = headers[i]  # O(1)
+                d[key] = data[i]  # O(1)
+            result.append(d)  # O(1)
+    return result
+    # O(1) + O(1) + O(1) + O(1) + (N - 1) * O(1) + (N - 1) * O(1) + (N - 1) * (M * O(1) + M * O(1)) + (N - 1) * O(1)
+    # (N - 1) * (3 * O(1) + (M * O(1) + M * O(1))
+    # -> (N - 1) * 2M * O(1) - > (N - 1) * 2M -> N * M -> O(N * M)
+    # -> O(N)
+
+def num_older_than(age_old:Union[int, float], dataname):
```
Comment (patrickkwang) [inline ehr_analysis2.py:23]: Please run `black`.

File: ehr_analysis2.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    for record in dataname: # N
+        birth = record["PatientDateOfBirth"]# 0(1)
+        b_str = datetime.strptime(birth, "%Y-%m-%d %H:%M:%S.%f")# 0(1)
+        age_new = (datetime.today() - b_str) / timedelta(days=365.2425)# 0(1)
+        if age_new > age_old:# 0(1)
+            num += 1# 0(1)
+    return num
+# 5 * N -> 0(N)
+
+def sick_patients(lab, gt_lt, value, lablist):
```
Comment (patrickkwang) [inline ehr_analysis2.py:34]: Does this function return something?

File: ehr_analysis2.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    @param patients: input data
+    @param patient_id: specific patient
+    @return: age of the specific patient
+    '''
+    date0 = datetime.now()
+    for patient in patients:
+        birth = datetime.strptime(patient[2], '%Y-%m-%d %H:%M:%S.%f')
+        if patient[0] == patient_id:
+            return round((date0 - birth).days / 365, 1)
+            break
```
Comment (patrickkwang) [inline ehr_analysis2.py:74]: This line is unreachable. You can remove it.

**Review by patrickkwang — CHANGES_REQUESTED** [review]
Did you make a new file here instead of editing the old one? It would be nice to be able to view only the changes here: https://github.com/JaneDeng28/Biostat821-EHR-analysis/pull/2/files

**Comment by JaneDeng28** [comment]
I did rebase using a git client and I just do rebase this branch to main, and keep the version of this file in branch "part-1". So I actually did not make any change to it since the last time you ask me to rebase it to main.

File: ehr_analysis2.py [hunk 5]
```diff
@@ -2,7 +2,7 @@
 from typing import Union
 
 """Data Parsing"""
-def parse_data(filename):
+def parse_data(filename:str) -> list:
```
Comment (patrickkwang) [inline ehr_analysis2.py:5]: List of what? Make type hints as specific as possible.

**Review by patrickkwang — CHANGES_REQUESTED** [review]
Please address all prior comments. Rather than this new `ehr_analysis2.py`, we want to edit the existing `ehr_analysis.py`. This is a "cumulative" project: each part is meant to be a modification/improvement of the last.

File: ehr_analysis2.py (hunk 5 above)
Comment (JaneDeng28) [reply ehr_analysis2.py:5]: ehr_analysis2.py is just a rename of ehr_analysis.py, when I do rebase, the name changed automatically. This is not a new version.

File: ehr_analysis.py [hunk 6]
```diff
@@ -20,25 +22,68 @@ def parse_data(filename):
...
-def sick_patients(lab, gt_lt, value, lab_name):
-    stat = lambda lab_value, ref_value: (gt_lt[0] == '>' and lab_value > ref_value) or\
-                                        (gt_lt[0] == '<' and lab_value < ref_value) # 0(1)
-    sick_list = [row[0] for row in labs[1:] if row[2] == lab_name and stat(float(row[3]), value)] # N * 0(1)
-    sick_list = list(set(sick_list)) # 0(1)
-    return sick_list
-# 0(1) + N * 0(1) + 0(1)
-# 0(N)
+
+def sick_patients(lab, gt_lt, value, lablist) -> list:
```
Comment (patrickkwang) [inline ehr_analysis.py:40]: Every input and output should have a type hint, and the type hints should be as specific as possible, e.g. `list[str]`. Please scan the entire file and fix these issues.

File: test_ehr_analysis.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    """
+    assert [
+        "64182B95-EB72-4E2B-BE77-8050B71498CE",
+        "Male",
+        "1952-01-18 19:51:12.917",
+        "African American",
+        "Separated",
+        "English",
+        "13.03",
+    ] in ehr_analysis.filename
```
Comment (patrickkwang) [inline test_ehr_analysis.py:17]: `ehr_analysis.filename` does not exist. One should be able to run `pytest test_ehr_analysis.py` after cloning the repo.

**Review by patrickkwang — CHANGES_REQUESTED** [review]

File: test_ehr_analysis.py (hunk 7 above)
Comment (patrickkwang) [reply test_ehr_analysis.py]: The tests still do not run. There is no `ehr_analysis.patient`, among other issues..

File: ehr_analysis.py (hunk 6 above)
Comment (patrickkwang) [reply ehr_analysis.py]: Please scan the entire file and fix all of the type hints.

**Review by patrickkwang — CHANGES_REQUESTED** [review]
Please address all comments.

**Pull request merged**